util/fmtutil/format.go

```diff
@@ -201,3 +199,102 @@ func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]stri
 
 	return labels
 }
+
+func makeBucketSpans(spans []*dto.BucketSpan) []prompb.BucketSpan {
+	out := make([]prompb.BucketSpan, 0, len(spans))
+	for _, span := range spans {
+		out = append(out, prompb.BucketSpan{Offset: span.GetOffset(), Length: span.GetLength()})
+	}
+	return out
+}
+
+func addClassicHistogramToWriteRequest(wr *prompb.WriteRequest, h *dto.Histogram, labels map[string]string, timestamp int64) {
+	metricName := labels[model.MetricNameLabel]
+
+	// Preserve metric name order with first bucket suffix timeseries then
+	// count suffix timeseries and finally sum suffix timeserie
+
+	// Add histogram bucket timeseries.
+	bucketLabels := make(map[string]string, len(labels)+1)
+	for key, value := range labels {
+		bucketLabels[key] = value
+	}
+	for _, b := range h.Bucket {
+		bucketLabels[model.MetricNameLabel] = metricName + bucketStr
+		bucketLabels[model.BucketLabel] = fmt.Sprint(b.GetUpperBound())
+		toTimeseries(wr, bucketLabels, timestamp, float64(b.GetCumulativeCount()))
+	}
+
+	// Overwrite label model.MetricNameLabel for count and sum metrics.
+
+	// Add histogram count timeseries.
+	labels[model.MetricNameLabel] = metricName + countStr
+	toTimeseries(wr, labels, timestamp, float64(h.GetSampleCount()))
+
+	// Add histogram sum timeseries.
+	labels[model.MetricNameLabel] = metricName + sumStr
+	toTimeseries(wr, labels, timestamp, h.GetSampleSum())
+}
+
+func addNativeHistogramToWriteRequest(wr *prompb.WriteRequest, h *dto.Histogram, labels map[string]string, timestamp int64) {
+	// Note that for float histograms:
+	// - Count should use CountFloat
+	// - NegativeCounts should bet used instead of NegativeDeltas
+	// - PositiveCounts should bet used instead of PositiveDeltas
+	//
+	// For float histograms:
+	// - Count should use CountInt
+	// - NegativeDeltas should bet used instead of NegativeCounts
+	// - PositiveDeltas should bet used instead of PositiveCounts
+	//
+	// *Counts and *Deltas are fine because the functions return nil if
+	// these are not set.
+	//
+	// The necessary interfaces for Count and ZeroCount are impossible to
+	// implement because they require a private method, this means it's not
+	// possible to create makeCount function that returns an instance of
+	// that interface.
+
+	hist := prompb.Histogram{
+		Sum:            h.GetSampleSum(),
+		Schema:         h.GetSchema(),
+		ZeroThreshold:  h.GetZeroThreshold(),
+		NegativeSpans:  makeBucketSpans(h.GetNegativeSpan()),
+		NegativeDeltas: h.GetNegativeDelta(),
+		NegativeCounts: h.GetNegativeCount(),
+		PositiveSpans:  makeBucketSpans(h.GetPositiveSpan()),
+		PositiveDeltas: h.GetPositiveDelta(),
+		PositiveCounts: h.GetPositiveCount(),
+		Timestamp:      timestamp,
+		// ResetHint: what is this?
```

Comment:
If the remote-write request is created from data in the TSDB, you can populate the ResetHint. But if you just have a single dto protobuf, you have to leave it as "unknown" (which is the default, I hope).